src/foundry/client/canvas/placeables/tokens/ruler.d.mts

```diff
@@ -1,13 +1,211 @@
-import type { Identity } from "#utils";
+import type { DeepReadonly, Identity } from "#utils";
 import type { BaseTokenRuler } from "./_module.d.mts";
 
 declare class TokenRuler extends BaseTokenRuler {
   #TokenRuler: true;
+
+  /**
+   * A handlebars template used to render each waypoint label.
+   */
+  static WAYPOINT_LABEL_TEMPLATE: string;
+
+  protected override _onVisibleChange(): void;
+
+  /**
+   * Configure the properties of the outline.
+   * Called in {@link TokenRuler#draw}.
+   * @returns The thickness in pixels and the color
+   */
+  protected _configureOutline(): TokenRuler.Outline;
+
+  /**
+   * Configure the properties of the dash line.
+   * Called in {@link TokenRuler#draw}.
+   * @returns The dash in pixels, the gap in pixels, and the speed in pixels per second
+   */
+  protected _configureDashLine(): TokenRuler.DashLine;
+
+  draw(): Promise<void>;
+
+  clear(): void;
+
+  destroy(): void;
+
+  refresh(rulerData: TokenRuler.Data): void;
+
+  /**
+   * Get the context used to render a ruler waypoint label.
+   */
+  protected _getWaypointLabelContext(
+    waypoint: TokenRuler.Waypoint,
+    state: TokenRuler.State,
+  ): TokenRuler.WaypointContext | void;
+
+  /**
+   * Get the style of the waypoint at the given waypoint.
+   * @param waypoint - The waypoint
+   * @returns  The radius, color, and alpha of the waypoint. If the radius is 0, no waypoint marker is drawn.
+   */
+  protected _getWaypointStyle(waypoint: TokenRuler.Waypoint): TokenRuler.WaypointStyle;
+
+  /**
+   * Get the style of the segment from the previous to the given waypoint.
+   * @param waypoint - The waypoint
+   * @returns The line width, color, and alpha of the segment.  If the width is 0, no segment is drawn.
+   */
+  protected _getSegmentStyle(waypoint: TokenRuler.Waypoint): TokenRuler.SegmentStyle;
+
+  /**
+   * Get the style to be used to highlight the grid offset.
+   * @param waypoint - The waypoint
+   * @param offset - An occupied grid offset at the given waypoint that is to be highlighted
+   * @returns The color, alpha, texture, and texture matrix to be used to draw the grid space.
+   *          If the alpha is 0, the grid space is not highlighted.
+   */
+  protected _getGridHighlightStyle(
+    waypoint: TokenRuler.Waypoint,
+    offset: DeepReadonly<foundry.grid.BaseGrid.Offset3D>,
+  ): TokenRuler.GridHighlightStyle;
 }
 
 declare namespace TokenRuler {
   interface Any extends AnyTokenRuler {}
   interface AnyConstructor extends Identity<typeof AnyTokenRuler> {}
+
+  interface Outline {
+    thickness: number;
+    color: PIXI.ColorSource;
+  }
+
+  interface DashLine {
+    dash: number;
+    gap: number;
+    speed: number;
+  }
+
+  interface Data {
+    /**
+     * The waypoints that were already passed by the Token
+     */
+    passedWaypoints: unknown[]; // TODO: Token.MeasuredMovementWaypoint
```

Comment:
Can you move the `unknown` inside of `Token`? I don't want someone to have to open this file to find the stubs.